File: signs.cpp
/*
Solution to signs UCFHSPT2005

@author yiuyuho@yahoo.com
@date 2005-05-07
*/

#include <iostream>
#include <fstream>
#include <cmath>

using namespace std;

const double esp=0.01,PI=3.14159;

//equal function to compare floating point valus, 
//in order to avoid precision error
bool eq(double x,double y)
{
	return fabs(x-y)<esp;
}

//Does sin(A) where A is in degrees
double dSin(double A)
{
	return sin(A*PI/180);
}

//convert radian to degree
double deg(double R)
{
	return R*180/PI;
}

//Use angle sum formula to solve for an unknown angle, if the other 2 angles are known
void angleSum(double &A,double &B,double &C)
{
	if(A>0 && B>0 && C<0) C=180-A-B;
	else if(A>0 && C>0 && B<0) B=180-A-C;
	else if(B>0 && C>0 && A<0) A=180-B-C;
}

//use sin(sign) law to solve for one of the A,a,B,b
//given the other 3 are known
bool sinLaw(double &A,double &a,double &B,double &b)
{
	if(A>0 && a>0 && B>0 && b<0) b=a*dSin(B)/dSin(A);
	else if(B>0 && b>0 && A>0 && a<0) a=b*dSin(A)/dSin(B);
	else if(A>0 && a>0 && b>0 && B<0) B=deg(asin(b*dSin(A)/a));
	else if(B>0 && b>0 && a>0 && A<0) A=deg(asin(a*dSin(B)/b));
	else return 0; //no new information is found
	//some new information is found
	return 1;
}

bool invalidSin(double A,double a,double B,double b)
{
	double sin;
	if(A>0 && a>0 && b>0 && B<0)
	{
		sin=b*dSin(A)/a;
		if(sin<0 || sin>1) return 1;
	}
	if(B>0 && b>0 && a>0 && A<0)
	{
		sin=a*dSin(B)/b;
		if(sin<0 || sin>1) return 1;
	}
	return 0;
}

//checks to see if any measurement is missing
double miss(double A,double B,double C,double a,double b,double c)
{
	return A<0 || B<0 || C<0 || a<0 || b<0 || c<0;
}

//checks that all angles are acute and their sum is in fact 180.
bool checkAng(double A,double B,double C)
{
	return eq(A+B+C,180) && 0<A && A<90 && 0<B && B<90 && 0<C && C<90;
}

//checks that A,a,B,b obeys sin law
bool checkSin(double A,double a,double B,double b)
{
	return eq(a*dSin(B),b*dSin(A));
}

//check for ending condition (all negative)
bool end(double A,double B,double C,double a,double b,double c)
{
	return A<0 && B<0 && C<0 && a<0 && b<0 && c<0;
}

int main()
{
	ifstream cin("signs.in");
	double A,B,C,a,b,c;
	bool done;
	int cc;
	cc=1;
	//set output precision
	cout.precision(2);
	cout.setf(ios::fixed|ios::showpoint);
	//input
	cin>>A>>B>>C>>a>>b>>c;
	while(!end(A,B,C,a,b,c))
	{
		cout<<"Sign #"<<cc++<<": ";
		done=0;
		while(!done)
		{
			done=1;
			angleSum(A,B,C);
			//if we find any new information, then we might want to loop
			//back because the state of data has changed, thus set done=0;
			if(invalidSin(A,a,B,b) || invalidSin(A,a,C,c) || invalidSin(B,b,C,c))
			{
			 goto reject;
			}
			if(sinLaw(A,a,B,b)) done=0;
			if(sinLaw(A,a,C,c)) done=0;
			if(sinLaw(B,b,C,c)) done=0;
		}
		//if some measurement still missing, say so
		if(miss(A,a,B,b,C,c))
		{
			cout<<"Lost!"<<endl;
		}
		else if(checkAng(A,B,C) && checkSin(A,a,B,b) && checkSin(A,a,C,c))
		{
			//everything passed angle and sin law, thus output
			cout<<A<<' '<<B<<' '<<C<<' '<<a<<' '<<b<<' '<<c<<endl;
		}
		else
		{
			//some of the measurement did not pass the checking
reject:
			cout<<"Rejected!"<<endl;
		}
		cin>>A>>B>>C>>a>>b>>c;
	}
	return 0;
}

File: signs.java
import java.io.*;
import java.util.*;

public class signs
{
/************************************
 * MAIN
 ************************************/
	 //Throw the IOException because it is unimportant and only an issue
	 //if I handle the input incorrectly or the file doesn't exist
	 public static void main(String[] args) throws IOException
	 {
	 	//create an instance of the problem solving class
	 	signs s = new signs();
	 	
	 	//run the problem solving solution
	 	s.run();
	 }
 
/************************************
 * Class Variables
 ************************************/
 	private BufferedReader in;
 	private double PI = 3.14159;


/************************************
 * Class Funtions
 ************************************/

 	//Constructor for the class
	//Throw the IOException the function creating this
	//so it doesn't clutter code here
    public signs() throws IOException
    {
    	in = new BufferedReader(new FileReader("signs.in"));
    }
 
 	public void run() throws IOException
 	{
 		/* Rules of the problem
		 *  Use the A + B + C == 180 Rule
		 *  Use a / sin A = b / sin B = c / sin C
		 *  All angles should be less than 90
		 *  0.01 inch tolerance
		 */
		 
		/* Input Specs
		 * Each line will contain Angle Angle Angle side side side
		 *  Angles will not be between 0 and 90
		 *  if measurement is negative then it is missing
		 *  a line with all negative measurements will indicate end of problem
		 */ 
		 
		 int Counter = 1;
		 do
		 {
		 	//Read the input line and tokenize it by spaces
		 	StringTokenizer st = new StringTokenizer(in.readLine(), " ");
		 	
		 	//Grab all the measurements
		 	double[] measurements = new double[6];
		 	boolean[] empty = new boolean[6];
		 	
		 	for(int i = 0; i < 6; i++)
		 	{
		 		measurements[i] = Double.parseDouble(st.nextToken());
		 		empty[i] = measurements[i] < 0.0;
		 	}
		 	
		 	//If all of them are empty(negative) then exit as this is the terminating case
		 	if(empty[0] && empty[1] && empty[2] && empty[3] && empty[4] && empty[5])
		 	{
		 		return;
		 	}
		 	
		 	//See if you can fill out the angles real quick
		 	for(int i = 0; i < 3; i++)
 			{
 				int A = i;
 				int B = (i + 1) % 3;
 				int C = (i + 2) % 3;
 				if(!empty[A] && !empty[B] && empty[C])
 				{
 					measurements[C] = 180.0 - measurements[A] - measurements[B];
 					empty[C] = false;
 			
 				}
 				else if(!empty[A] && empty[B] && !empty[C])
 				{
 					empty[B] = false;
 					measurements[B] = 180.0 - measurements[A] - measurements[C];
 			
 				}
 			}
		 	
		 	//You wont be able to figure it out if you don't have a pair, so check
		 	int pair = -1;
		 	
		 	for(int i = 0; i < 3; i++)
		 	{
		 		if(!empty[i] && !empty[i+3])
		 		{
		 			pair = i;
		 			break;
		 		}
		 	}
		 	
		 	int emptyPairs = 0;
		 		
		 	//Won't be able to figure it out if a two groupings are empty either
	 		for(int i = 0; i < 3; i++)
	 		{
	 			if(empty[i] && empty[i + 3])
	 			{
	 				emptyPairs++;
	 			}
	 		}

	 		

			//If solvable then try to solve it	 		
		 	if(pair != -1 && emptyPairs <= 1)
		 	{
		 		boolean changed = false;
		 		
		 		//I didn't want to write a bunch of if's and loops so 
		 		//I just wrote two.
		 		//While the values are changing it'll keep doing the following
		 		//Checking to see if it can solve for a value
		 		//Solve for it
		 		//Solve for a missing angle
		 		do
		 		{
		 			changed = false;
		 			for(int i = 0; i < 3; i++)
		 			{
		 				//Skip over pair because we already know it's got both valueas
		 				if(i != pair)
		 				{
		 					
		 					//If the angle is missing but we have the side
		 					if(empty[i] && !empty[i + 3])
		 					{
		 						empty[i] = false;
		 						int side = i + 3;
		 						int angle = i;
		 						
		 						//Applying the law given in the problem spec
		 						//You have to multiply the angle by PI/180.0 to get it from degrees to radians
		 						//Multiply by 180.0 / PI to change it back to degrees
		 						double sinValue = measurements[side] * Math.sin((measurements[pair] * PI)/180.0);
		 						sinValue /= measurements[pair + 3];
		 						measurements[angle] = (Math.asin(sinValue) * 180.0) / PI;
		 						changed = true;
		 					}
		 					//If the side is missing but we have the angle
		 					else if(!empty[i] && empty[i + 3])
		 					{
		 						empty[i + 3] = false;
		 						
		 						int side = i + 3;
		 						int angle = i;
		 						
		 						//Applying the law given in the problem spec
		 						//You have to multiply the angle by PI/180.0 to get it from degrees to radians
		 						double sideValue = Math.sin(measurements[angle] * PI/180.0) * measurements[pair+3];
		 						sideValue /= Math.sin(measurements[pair] * PI/180.0);
		 						measurements[side] = sideValue;
		 						changed = true;
		 					}
		 				}
		 			}
		 			
		 			//Check the angles and fill in the blanks if possible
		 			//This uses the A + B + C = 180 rule
		 			for(int i = 0; i < 3; i++)
		 			{
		 				int A = i;
		 				int B = (i + 1) % 3;
		 				int C = (i + 2) % 3;
		 				if(!empty[A] && !empty[B] && empty[C])
		 				{
		 					measurements[C] = 180.0 - measurements[A] - measurements[B];
		 					empty[C] = false;
		 					changed = true;		 					
		 				}
		 				else if(!empty[A] && empty[B] && !empty[C])
		 				{
		 					empty[B] = false;
		 					measurements[B] = 180.0 - measurements[A] - measurements[C];
		 					changed = true;
		 				}
		 			}
		 		}while(changed);


				//Make sure it fits to all the rules
				//First Signs
				boolean lawOfSigns = true;
				
				//If the values for law of signs are within the tolerance of each other then the 
				//data abides by the lawOfSigns else it marks that it doesn't
				for(int i = 0; i < 3; i++)
				{
					double lawValue = measurements[i + 3] / Math.sin(measurements[i] * PI/180.0);

					for(int j = 0; j < 3; j++)
					{
						if(i == j) continue;

						double temp = measurements[j + 3] / Math.sin(measurements[j ] * PI/180.0);
						
						if(!equal(temp,lawValue))
						{
							lawOfSigns = false;
						}
					}
				}
				
				//Do the angles add up to 180 degrees?
				boolean angles = equal(measurements[0] + measurements[1] + measurements[2],180.0);

                                angles = angles && measurements[0] < 90.0 && measurements[1] < 90.0 && measurements[2] < 90.0;
				
				//Print the header
				System.out.print("Sign #" + Counter + ": ");
				
				//If the triangle passed print the data
				if(lawOfSigns && angles)
				{
			 		for (int i = 0; i < 6; i++)
			 		{
			 			System.out.print(format(measurements[i]) + " ");
			 		}
		 		}
		 		//Else print rejected
		 		else
		 		{
		 			System.out.print("Rejected!");
				
		 		}
		 		System.out.println("");
		 	}
		 	else
		 	{
		 		System.out.println("Sign #" + Counter + ": Lost!");
		 	}
		 	
		 	Counter++;

		 }while(true);
 	}
 	
 	private String format(double d)
 	{
 		//Since we are only rounding to two decimal places we can add 0.005 to do automatic rounding
 		d += 0.005;
 		
 		//make it a string
 		String value = d + "";
 		
 		//remove the excess garbage that we aren't using
 		//should return a number formatted to two decimal places
 		return value.substring(0, value.indexOf(".") + 3);
 	}
 	
 	private boolean equal(double a, double b)
 	{
 		//if the absolute difference between the two values is less than
 		//the tolerance then the values are the same
 		return Math.abs(a-b) <= 0.01	;
 	}
}
